static/app/components/compactSelect/listBox.tsx

```diff
@@ -288,12 +311,98 @@ function ListBox<Value extends React.Key>({
             );
           })}
       </SelectListBoxWrap>
+
+      {multiple &&
+        filteredItems.map(item => {
+          if (item.type !== 'section') {
+            return null;
+          }
+
+          return (
+            <HiddenSectionToggle
+              key={item.key}
+              id={`${listBoxProps.id}-section-toggle-${item.key}`}
+              item={item}
+              listState={listState}
+              listBoxRef={ref}
+            />
+          );
+        })}
     </Fragment>
   );
 }
 
 export {ListBox};
 
+interface HiddenSectionToggleProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
+  item: Node<any>;
+  listBoxRef: React.RefObject<HTMLUListElement>;
+  listState: ListState<any>;
+}
+
+/**
+ * A visually hidden but keyboard-focusable button to toggle (select/unselect) all
+ * options in a given section. We need these hidden buttons because the visible toggle
+ * buttons inside ListBox are not keyboard-focusable (due to ListBox implementing roving
+ * `tabindex`).
+ */
+function HiddenSectionToggle({
+  item,
+  listState,
+  listBoxRef,
+  ...props
+}: HiddenSectionToggleProps) {
+  // Highlight this toggle's visible counterpart (rendered inside the list box) on focus
+  const {focusProps} = useFocus({
+    onFocus: () => {
+      const visibleCounterpart = listBoxRef.current?.querySelector(
+        `li[data-key="${item.key}"] button[aria-hidden]`
+      );
+
+      if (!visibleCounterpart) {
+        return;
+      }
+      visibleCounterpart.classList.add('focus-visible');
+    },
+    onBlur: () => {
+      const visibleCounterpart = listBoxRef.current?.querySelector(
+        `li[data-key="${item.key}"] button[aria-hidden]`
+      );
+
+      if (!visibleCounterpart) {
+        return;
+      }
+      visibleCounterpart.classList.remove('focus-visible');
+    },
+  });
```

Comment:
The result:
https://user-images.githubusercontent.com/44172267/218596994-283446e5-10f1-44fb-a78d-2dd49ee7c721.mov